added feature to propagate function signs to all cells and its accord…

Jurwen opened the pull request:
Added Feature to Propagate Function Signs to all Cells

* Added `sign_propogation` function, and changed its related API and IO stream to output `cell_function label`.
* Added four configurations of three spheres model to the unit tests. 
* Adjusted other units to test for the new feature.



<img src=https://github.com/duxingyi-charles/Robust-Implicit-Surface-Networks/assets/46691179/59d26ab2-b47a-48fe-a39e-947e47957d9f height="300">



Visual verifications of the 4 configurations of three spheres. On each plot, there is an overlay of overall mesh and a cell. The cell is plot in color, and on top of it, there are TF labels of all three spherical functions for this cell. For the label, 1 means inside the function, and 0 means outside.

qnzhou commented on tests/test_implicit_networks.cpp:
```diff
@@ -187,20 +200,190 @@ TEST_CASE("implicit arrangement on known examples", "[IA][examples]") {
+        REQUIRE(patch_function_label == patch_gt);
+        std::vector<std::vector<size_t>> cell_gt = {{0, 0, 1}, {0, 0, 0}, {0, 1, 1}, {1, 1, 1}};
+        REQUIRE(cell_function_label == cell_gt);
     }
```
Can you add another 3 sphere unit test: The first sphere has very large radius such that its isosurface does not intersect the grid. The second sphere and third sphere are regular sized sphere intersecting each other. The second sphere has normal orienting outward, and the third sphere has normal orienting inward.

qnzhou replied:
> The function could have an oversized shape that are not captured by the bounding, like this example, or a highly curved shape such that the grid is too coarse to capture the function. The easy solution to this issue is that once we find a function that its sign is not activated by any of the patches, we can propagate its sign at any point on the grid to all the cells. Would that be too hand-wavy? Or is there a better solution to this issue?

Yes, this test case is designed to check this situation. If an isosurface does not touch the grid at all, we can assume it has the same sign for all cells based on our assumptions. Since we already have the sign of the function at grid points, we can use that as the sign for all cells.

Jurwen commented:
* Updated data type (size_t --> bool) and code style (removed comments)
* Added example of three spheres. For any function that does not appear (activate) in the bounding box, the default value for all the cells' labels of this function is currently set to be `false`. 

## Here is an updated overview of 3-sphere examples.

<img src=https://github.com/duxingyi-charles/Robust-Implicit-Surface-Networks/assets/46691179/fe57b2d9-541e-4f24-92ff-789e3d5bde1c height="500">

qnzhou commented on tests/test_implicit_networks.cpp:
```diff
@@ -187,21 +200,231 @@ TEST_CASE("implicit arrangement on known examples", "[IA][examples]") {
+        funcVals.resize(n_pts, n_func);
+        size_t func_id;
+        
+        if (!load_functions(std::string(TEST_FILE) + "/3-sphere-5.json", pts, funcVals)) {
```
Please check in the test files.

Jurwen replied:
Added test files. Previously blocked by .gitignore.

qnzhou reviewed (approved):
Thanks Owen! Looks good to me!
